### PR TITLE
core/rawdb: use atomic int to prevent data race

### PR DESCRIPTION
**Pull request by andicrypt**
Currently, some fields (i.e., `frozen`) in rawdb packet are currently stored atomically but read non-atomically, causing a data race. As a result, one of our shadow nodes (with data race detector flag set) meets a race condition error.
```
WARNING: DATA RACE
Write at 0x00c0001fa280 by main goroutine:
  sync/atomic.StoreInt64()
      runtime/race_amd64.s:237 +0xb
  sync/atomic.StoreUint64()
      <autogenerated>:1 +0x1b
  github.com/ethereum/go-ethereum/core/rawdb.(*freezerdb).TruncateAncients()
      <autogenerated>:1 +0x67
  github.com/ethereum/go-ethereum/node.(*closeTrackingDB).TruncateAncients()
      <autogenerated>:1 +0x56
  github.com/ethereum/go-ethereum/core.(*BlockChain).setHeadBeyondRoot.func2()
      github.com/ethereum/go-ethereum/core/blockchain.go:728 +0xbb
  github.com/ethereum/go-ethereum/core.(*HeaderChain).SetHead()
      github.com/ethereum/go-ethereum/core/headerchain.go:598 +0xb72
  github.com/ethereum/go-ethereum/core.(*BlockChain).setHeadBeyondRoot()
      github.com/ethereum/go-ethereum/core/blockchain.go:746 +0x33b
  github.com/ethereum/go-ethereum/core.NewBlockChain()
      github.com/ethereum/go-ethereum/core/blockchain.go:335 +0x1d8e
...
...
Previous read at 0x00c0001fa280 by goroutine 4197:
  github.com/ethereum/go-ethereum/core/rawdb.(*freezer).freeze()
      github.com/ethereum/go-ethereum/core/rawdb/freezer.go:384 +0x553
  github.com/ethereum/go-ethereum/core/rawdb.NewDatabaseWithFreezer.func1()
      github.com/ethereum/go-ethereum/core/rawdb/database.go:223 +0x4c

Goroutine 4197 (running) created at:
  github.com/ethereum/go-ethereum/core/rawdb.NewDatabaseWithFreezer()
      github.com/ethereum/go-ethereum/core/rawdb/database.go:222 +0xeb9
  github.com/ethereum/go-ethereum/core/rawdb.Open()
      github.com/ethereum/go-ethereum/core/rawdb/database.go:346 +0xc8
```

To mitigate this, this PR migrates to using atomic.Uint64 for some fields to enforce developer read/write these fields in atomic access manner.
Full log: [full_log.txt](https://github.com/user-attachments/files/16747300/untitled2.txt)

Reference: https://github.com/ethereum/go-ethereum/pull/26935/files

### CODE REVIEW

File: core/rawdb/freezer.go [hunk 1]
```diff
@@ -446,8 +443,8 @@ func (f *freezer) freeze(db ethdb.KeyValueStore) {
 		batch.Reset()
 
 		// Step into the future and delete and dangling side chains
-		if f.frozen > 0 {
-			tip := f.frozen
+		if f.frozen.Load() > 0 {
+			tip := f.frozen.Load()
```
Comment (minh-bq) [inline core/rawdb/freezer.go:447]: nit:
```go
tip := f.frozen.Load()
if tip > 0 {
```

**Review by minh-bq — COMMENTED** [review]
The description is quite misleading. The old way (i.e. using atomic.StoreUint64) to perform atomic read/write is still correct. The data race occurs because some reads are not atomic. IMO, the new way help to reduce programming error as it forces developer to using atomic read/write to get the underlying value.

**Comment by andicrypt** [comment]
> The description is quite misleading. The old way (i.e. using atomic.StoreUint64) to perform atomic read/write is still correct. The data race occurs because some reads are not atomic. IMO, the new way help to reduce programming error as it forces developer to using atomic read/write to get the underlying value.
Fixed description!

**Comment by andicrypt** [comment]
> The description is quite misleading. The old way (i.e. using atomic.StoreUint64) to perform atomic read/write is still correct. The data race occurs because some reads are not atomic. IMO, the new way help to reduce programming error as it forces developer to using atomic read/write to get the underlying value.

Fixed description

File: core/rawdb/freezer_table.go [hunk 2]
```diff
@@ -89,10 +89,8 @@ func (start *indexEntry) bounds(end *indexEntry) (startOffset, endOffset, fileId
 // It consists of a data file (snappy encoded arbitrary data blobs) and an indexEntry
 // file (uncompressed 64 bit indices into the data file).
 type freezerTable struct {
-	// WARNING: The `items` field is accessed atomically. On 32 bit platforms, only
-	// 64-bit aligned fields can be atomic. The struct is guaranteed to be so aligned,
-	// so take advantage of that (https://golang.org/pkg/sync/atomic/#pkg-note-BUG).
-	items uint64 // Number of items stored in the table (including items removed from tail)
+	items      atomic.Uint64 // Number of items stored in the table (including items removed from tail)
+	itemOffset atomic.Uint64 // Number of items removed from the table
```
Comment (minh-bq) [inline core/rawdb/freezer_table.go:93]: Why do we change itemOffset from 32-bit to 64-bit?

Comment (andicrypt) [reply core/rawdb/freezer_table.go:93]: As I see that type cast itemOffset to uint64 frequently happens and uint64 might be good for x64 machine when handle atomic read/write

**Review by minh-bq — APPROVED** [review]

**Pull request merged**